Model/Api/OAuthRedirect.php

```diff
@@ -35,37 +46,199 @@ class OAuthRedirect implements OAuthRedirectInterface
+    public function login($code = '', $scope = '', $state = '')
     {
         if (!$this->deciderHelper->isBoltSSOEnabled()) {
             throw new NoSuchEntityException(__('Request does not match any route.'));
         }
 
-        $this->response->setHeader('Content-Type', 'application/json');
-        $this->response->setHttpResponseCode(500);
-        $this->response->setBody(json_encode(['message' => 'Not implemented']));
-        $this->response->sendResponse();
+        if ($code === '' || $scope === '' || $state === '') {
+            throw new WebapiException(__('Bad Request'), 0, WebapiException::HTTP_BAD_REQUEST);
+        }
+
+        list($clientID, $clientSecret, $boltPublicKey) = $this->ssoHelper->getOAuthConfiguration();
+
+        $token = $this->ssoHelper->exchangeToken($code, $scope, $clientID, $clientSecret);
+        if ($token === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $payload = $this->ssoHelper->parseAndValidateJWT($token, $clientID, $boltPublicKey);
+        if ($payload === null) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
+        }
+
+        $externalID = $payload['sub'];
+
+        $websiteId = $this->storeManager->getStore()->getWebsiteId();
+        $storeId = $this->storeManager->getStore()->getId();
+
+        $externalCustomerEntity = null;
+        $customer = null;
+        try {
+            $externalCustomerEntity = $this->externalCustomerEntityRepository->getByExternalID($externalID);
+        } catch (NoSuchEntityException $nsee) {
+            // external customer entity not found
+        }
+        try {
+            if ($externalCustomerEntity !== null) {
+                $customer = $this->customerRepository->getById($externalCustomerEntity->getCustomerID());
+            } else {
+                $customer = $this->customerRepository->get($payload['email'], $websiteId);
+            }
+        } catch (NoSuchEntityException $nsee) {
+            // customer not found
+        }
+
+        // External customer entity exists, but the customer it links to doesn't
+        // This should happen very rarely, if at all, which is why we notify bugsnag when this happens
+        if ($externalCustomerEntity !== null && $customer === null) {
+            $this->bugsnag->notifyError(
+                'OAuthRedirect',
+                'external customer entity ' . $externalID . ' linked to nonexistent customer ' . $externalCustomerEntity->getCustomerID()
+            );
+        }
+
+        // If the customer isn't linked and it exists in M2, but the email is not verified, we throw an exception
+        if ($externalCustomerEntity === null && $customer !== null && !$payload['email_verified']) {
+            throw new WebapiException(__('Internal Server Error'), 0, WebapiException::HTTP_INTERNAL_ERROR);
```

Comment:
we definitely want to track how often this one happens-- do we have a way other than bugsnag?

Reply:
i don't think so. i think bugsnag is all we get.

Reply:
i'll add an error to bugsnag though

Reply:
We want to have the ability to write to datadog, but this task is not done yet